### ISSUE TITLE
9.0.1 beta1 in CentOS 6.6 doesn't honor share links after upgrade from 9.0.0 (shared links from 9.0.0 won't work)

### ISSUE THREAD

**Issue by jvillafanez** [owncloud/core#23512]
### Steps to reproduce
1. Install ownCloud 9.0.0 on CentOS 6.6
2. Upload a file, share it by link, and write down the generated link (generated link is something like "http://HOST:PORT/s/Mf40bDYW7gARbaI")
3. Verify the link works and anyone can download the file
4. Upgrade the server to 9.0.1 beta 1
5. After a successful upgrade, try to download the file through the link
### Expected behaviour

The file is downloaded
### Actual behaviour

File not found. Looks like the expected url is http://HOST:PORT/index.php/s/Mf40bDYW7gARbaI
### Server configuration

**Operating system**: CentOS 6.6

**Web server:** Apache/2.2.15

**Database:** mysql

**PHP version:** 5.4.16

**ownCloud version:** ownCloud 9.0.1 beta1

**Updated from an older ownCloud or fresh install:** upgrade from 9.0.0
### Notes
- Upgrades from earlier versions (8.2.3, and earlier) don't present this issue.
- After the upgrade to 9.0.1 (beta1), the file shows the correct link (including the "index.php" part) when you want the get the link from the files view. The link "token" (the "Mf40bDYW7gARbaI" in the url) is the same as before the upgrade.
- Ubuntu upgrades don't have this issue.

**Comment by LukasReschke** [comment]
Can you add more information on how you did update and install ownCloud? The index.php is not required if the `.htaccess` is writable at install and update time (both has to happen via web). If at 9.0.0 install time it's writable and at 9.0.1 it's not then all future URLs require `index.php` in it.

**Comment by rullzer** [comment]
Is the share still present in the oc_share table?

**Comment by jvillafanez** [comment]
@LukasReschke both install and update were done using tarballs. The upgrade was done by removing all contents inside the ownCloud installation (except the config and data folders) and unpacking the new version there, with the new .htaccess file. I guess the .htaccess file was overwritten (I think this is expected)

@rullzer The share is present, and the file can be downloaded by including the "index.php" part in the link

**Comment by LukasReschke** [comment]
> and unpacking the new version there, 

And then using the web interface or OCC do to the update?

> I guess the .htaccess file was overwritten (I think this is expected)

Please post your `.htaccess` file as well :smile:

**Comment by jvillafanez** [comment]
The upgrade was done using occ.

The contents of the last .htaccess file:

```
<IfModule mod_headers.c>
  <IfModule mod_setenvif.c>
    <IfModule mod_fcgid.c>
       SetEnvIfNoCase ^Authorization$ "(.+)" XAUTHORIZATION=$1
       RequestHeader set XAuthorization %{XAUTHORIZATION}e env=XAUTHORIZATION
    </IfModule>
    <IfModule mod_proxy_fcgi.c>
       SetEnvIfNoCase Authorization "(.+)" HTTP_AUTHORIZATION=$1
    </IfModule>
  </IfModule>

  <IfModule mod_env.c>
    # Add security and privacy related headers
    Header set X-Content-Type-Options "nosniff"
    Header set X-XSS-Protection "1; mode=block"
    Header set X-Robots-Tag "none"
    Header set X-Frame-Options "SAMEORIGIN"
    Header set X-Download-Options "noopen"
    Header set X-Permitted-Cross-Domain-Policies "none"
    SetEnv modHeadersAvailable true
  </IfModule>

  # Add cache control for CSS and JS files
  <FilesMatch "\.(css|js)$">
    Header set Cache-Control "max-age=7200, public"
  </FilesMatch>
</IfModule>
<IfModule mod_php5.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value always_populate_raw_post_data -1
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_php7.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_rewrite.c>
  RewriteEngine on
  RewriteRule .* - [env=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
  RewriteRule ^\.well-known/host-meta /public.php?service=host-meta [QSA,L]
  RewriteRule ^\.well-known/host-meta\.json /public.php?service=host-meta-json [QSA,L]
  RewriteRule ^\.well-known/carddav /remote.php/dav/ [R=301,L]
  RewriteRule ^\.well-known/caldav /remote.php/dav/ [R=301,L]
  RewriteRule ^remote/(.*) remote.php [QSA,L]
  RewriteRule ^(build|tests|config|lib|3rdparty|templates)/.* - [R=404,L]
  RewriteCond %{REQUEST_URI} !^/.well-known/acme-challenge/.*
  RewriteRule ^(\.|autotest|occ|issue|indie|db_|console).* - [R=404,L]

  # Rewrite rules for `front_controller_active`
  Options -MultiViews
  RewriteRule ^core/js/oc.js$ index.php [PT,E=PATH_INFO:$1]
  RewriteRule ^core/preview.png$ index.php [PT,E=PATH_INFO:$1]
  RewriteCond %{REQUEST_FILENAME} !\.(css|js|svg|gif|png|html|ttf|woff|ico)$
  RewriteCond %{REQUEST_FILENAME} !core/img/favicon.ico$
  RewriteCond %{REQUEST_FILENAME} !/remote.php
  RewriteCond %{REQUEST_FILENAME} !/public.php
  RewriteCond %{REQUEST_FILENAME} !/cron.php
  RewriteCond %{REQUEST_FILENAME} !/core/ajax/update.php
  RewriteCond %{REQUEST_FILENAME} !/status.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v1.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v2.php
  RewriteCond %{REQUEST_FILENAME} !/updater/
  RewriteCond %{REQUEST_FILENAME} !/ocs-provider/
  RewriteCond %{REQUEST_URI} !^/.well-known/acme-challenge/.*
</IfModule>
<IfModule mod_mime.c>
  AddType image/svg+xml svg svgz
  AddEncoding gzip svgz
</IfModule>
<IfModule mod_dir.c>
  DirectoryIndex index.php index.html
</IfModule>
AddDefaultCharset utf-8
Options -Indexes
<IfModule pagespeed_module>
  ModPagespeed Off
</IfModule>
```

**Comment by LukasReschke** [comment]
Then this is the expected behaviour. For the short URLs to work the .htaccess needs to be writable _AND_ the installation/update has to happen via web. The problem is that when using OCC we don't know the webroot which is required.

@nickvergessen @DeepDiver1975 Would it make sense to write "overwritewebroot" by default to the config file so we can also use it from OCC?

**Comment by nickvergessen** [comment]
overwritewebroot or overwrite.cli.url

**Comment by jvillafanez** [comment]
Do we have a solution for this? It's still happening

**Comment by cmonteroluque** [comment]
@jnweiger?

**Comment by shaneonabike** [comment]
This also happened to me where old shares were no longer working. Essentially, because the URL for gallery had changed to a generic share URL I believe.

I used the unix package update and then ran occ update. My htaccess is in fact allowed for group and user write access so I don't think that's the issue...

<code>
<IfModule mod_headers.c>
  <IfModule mod_setenvif.c>
    <IfModule mod_fcgid.c>
       SetEnvIfNoCase ^Authorization$ "(.+)" XAUTHORIZATION=$1
       RequestHeader set XAuthorization %{XAUTHORIZATION}e env=XAUTHORIZATION
    </IfModule>
    <IfModule mod_proxy_fcgi.c>
       SetEnvIfNoCase Authorization "(.+)" HTTP_AUTHORIZATION=$1
    </IfModule>
  </IfModule>
  <IfModule mod_env.c>
    # Add security and privacy related headers
    Header set X-Content-Type-Options "nosniff"
    Header set X-XSS-Protection "1; mode=block"
    Header set X-Robots-Tag "none"
    Header set X-Frame-Options "SAMEORIGIN"
    Header set X-Download-Options "noopen"
    Header set X-Permitted-Cross-Domain-Policies "none"
    SetEnv modHeadersAvailable true
  </IfModule>
  # Add cache control for CSS and JS files
  <FilesMatch "\.(css|js)$">
    Header set Cache-Control "max-age=7200, public"
  </FilesMatch>
</IfModule>
<IfModule mod_php5.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value always_populate_raw_post_data -1
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_php7.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_rewrite.c>
  RewriteEngine on
  RewriteRule .* - [env=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
  RewriteRule ^\.well-known/host-meta /public.php?service=host-meta [QSA,L]
  RewriteRule ^\.well-known/host-meta\.json /public.php?service=host-meta-json [QSA,L]
  RewriteRule ^\.well-known/carddav /remote.php/dav/ [R=301,L]
  RewriteRule ^\.well-known/caldav /remote.php/dav/ [R=301,L]
  RewriteRule ^remote/(.*) remote.php [QSA,L]
  RewriteRule ^(build|tests|config|lib|3rdparty|templates)/.* - [R=404,L]
  RewriteCond %{REQUEST_URI} !^/.well-known/acme-challenge/.*
  RewriteRule ^(\.|autotest|occ|issue|indie|db_|console).* - [R=404,L]
  # Rewrite rules for `front_controller_active`
  Options -MultiViews
  RewriteRule ^core/js/oc.js$ index.php [PT,E=PATH_INFO:$1]
  RewriteRule ^core/preview.png$ index.php [PT,E=PATH_INFO:$1]
  RewriteCond %{REQUEST_FILENAME} !\.(css|js|svg|gif|png|html|ttf|woff|ico|jpg|jpeg)$
  RewriteCond %{REQUEST_FILENAME} !core/img/favicon.ico$
  RewriteCond %{REQUEST_FILENAME} !/remote.php
  RewriteCond %{REQUEST_FILENAME} !/public.php
  RewriteCond %{REQUEST_FILENAME} !/cron.php
  RewriteCond %{REQUEST_FILENAME} !/core/ajax/update.php
  RewriteCond %{REQUEST_FILENAME} !/status.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v1.php
  RewriteCond %{REQUEST_FILENAME} !/ocs/v2.php
  RewriteCond %{REQUEST_FILENAME} !/updater/
  RewriteCond %{REQUEST_FILENAME} !/ocs-provider/
  RewriteCond %{REQUEST_URI} !^/.well-known/acme-challenge/.*
</IfModule>
<IfModule mod_mime.c>
  AddType image/svg+xml svg svgz
  AddEncoding gzip svgz
</IfModule>
<IfModule mod_dir.c>
  DirectoryIndex index.php index.html
</IfModule>
AddDefaultCharset utf-8
Options -Indexes
<IfModule pagespeed_module>
  ModPagespeed Off
</IfModule>
#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
ErrorDocument 403 /core/templates/403.php
ErrorDocument 404 /core/templates/404.php
<IfModule mod_rewrite.c>
  RewriteRule . index.php [PT,E=PATH_INFO:$1]
  RewriteBase /
  <IfModule mod_env.c>
    SetEnv front_controller_active true
    <IfModule mod_dir.c>
      DirectorySlash off
    </IfModule>
  </IfModule>
</IfModule>

</code>

**Comment by PVince81** [comment]
The same happened on my openSUSE 13.1 box using the RPM packages.
A recent update seems to have removed whichever settings were needed to strip "index.php".
Will have to redo this manually and already forgot how.

**Comment by SergioBertolinSG** [comment]
Still happening upgrading to 9.1 tarball.

**Comment by jnweiger** [comment]
@SergioBertolinSG do you have a reproducer? I'd like to log in and check out some differnt .htaccess settings.

**Comment by SergioBertolinSG** [comment]
@jnweiger It is an upgrade from 9.0.2 to 9.1 using tarballs, having public shared links before upgrading.

**Comment by PVince81** [comment]
@SergioBertolinSG when updating the tarball, did you replace .htaccess ?

**Comment by SergioBertolinSG** [comment]
Yes, it is replaced with the upgrade.

**Comment by PVince81** [comment]
@jnweiger and update ? Is there anything we can do to fix this ?

**Comment by SergioBertolinSG** [comment]
Observed in an upgrade from centos 6 CE 9.0.0 package to 9.1.0.8 CE package.

**Comment by PVince81** [comment]
@jnweiger I saw your documentation PR https://github.com/owncloud/documentation/pull/2489, is this related ?

**Comment by jnweiger** [comment]
@PVince81 not directly related. As we now support short urls without index.php,  we have changes in .htaccess and apache's owncloud.conf, probably this issue is collateral damage from these changes.

That is why I reviewed the options.

**Comment by PVince81** [comment]
@jnweiger any chance for a fix for this on the packaging side ?

**Comment by shaneonabike** [comment]
So is there any quick fix for this?

**Comment by PVince81** [comment]
Not sure, I guess you might need to redo the htaccess config manually to switch it back to the mode that has no "index.php"

**Comment by PVince81** [comment]
@SergioBertolinSG @jvillafanez can you check whether ".htaccess" is still writable for the web server user after the upgrade ?

I found in my local tests that if 9.0.2 was configured with removed "index.php" thanks to writable ".htaccess", then upgrading to 9.0.3 with a different but writable ".htaccess" will automatically restore that mode, see https://github.com/owncloud/documentation/issues/2205#issuecomment-229118822

So I suspect that at some point ".htaccess" was not writable. Or it was set writable too late, after "occ upgrade" already ran. @jnweiger can you double check the upgrade scripts ?

**Comment by SergioBertolinSG** [comment]
-rw-r--r--  1 root   apache  2492 Jun 16 16:55 .htaccess

I guess it is not.

**Comment by PVince81** [comment]
Without packaging, I redid some tests and this is what I observed:

1) Test 1: Setup 9.0.2 with writable `.htaccess`, upgrade to 9.0.3RC1 but **without** writable `.htaccess`
=> index.php doesn't work any more, however config.php does contain "htaccess.RewriteBase".
To make it work again, one needs to make `.htaccess` upgrade and trick OC into running the upgrade process by fiddling with versions (not recommended!)

2) Test 2: Setup 9.0.2 with writable `.htaccess`, upgrade to 9.0.3RC1 **with** writable `.htaccess`
=> index.php removal still works, config.php contains "htaccess.RewriteBase".

@jnweiger some proposals for a solution for distro packages:

A) Solution A: before during pre-setup (before replacing the sources), take note whether `.htaccess` is writable. After replacing the sources and before running "occ upgrade", set back the writable flag on `.htaccess` if applicable.

B) Solution B: before running "occ upgrade", grep config.php and see whether "htaccess.RewriteBase" is set. If yes, then make `.htaccess` writable. (note: RewriteBase only exists since 9.0.3!)

@jnweiger what do you think ?

**Comment by jnweiger** [comment]
As before, .htaccess should always be writable. Both solutions A) and B) seem to imply that.
This was only asserted for DEB packages. Added

**Comment by jnweiger** [comment]
@jvillafanez do you confirm, the share links now survive the upgrade, when using 9.0-3 packages?

**Comment by SergioBertolinSG** [comment]
This is working fine now.

**Comment by PVince81** [comment]
Hurrah! :cocktail:

**Comment by SergioBertolinSG** [comment]
While this problem was fixed using packages, it has come back in tarball upgrades from 9.0.X to 9.1.0RC1.

Seems like sharing API creates public shares in 9.0.X without index.php in the url and when downloading the files using requests(python library) it fails returning a 404 not found.

**Comment by PVince81** [comment]
@SergioBertolinSG so to clarify, are you saying that in 9.0.X the API creates public shares **without** index.php even when this kind of URL doesn't work due to the setup ?

If that's the case, I suspect that maybe OC doesn't properly recognize that this mode isn't supported.

**Comment by PVince81** [comment]
@SergioBertolinSG do you have a test setup with CentOS with 9.0.2 on which I could create such links ? When creating the links in the automated test, I guess it was using the OCS Share API and not the web UI ? (the web UI might still be correct)

**Comment by SergioBertolinSG** [comment]
Look at the original steps: 

> Upload a file, share it by link, and write down the generated link (generated link is something like "http://HOST:PORT/s/Mf40bDYW7gARbaI")

index.php is missing there.

**Comment by PVince81** [comment]
Things to check on the 9.0.2 CentOS instance:
- value of `oc_config.modRewriteWorking` in the web browser
- create link share from the OCS Share API: `curl 'http://admin:admin@localhost/owncloud/ocs/v2.php/apps/files_sharing/api/v1/shares?format=json' -H 'Content-Type: application/x-www-form-urlencoded; charset=UTF-8' -H 'Connection: keep-alive' --data 'shareType=3&permissions=15&path=%2Fwelcome.txt'`, does the link contain index.php ?

**Comment by PVince81** [comment]
Looks like it is reading the value "front_controller_active" to find out whether "index.php" part can be removed: https://github.com/owncloud/core/blob/v9.0.2/core/js/config.php#L153
and the server router does the same: https://github.com/owncloud/core/blob/v9.0.2/lib/private/route/router.php#L73

Maybe it's a false positive.

So I suspect that even though this value is true, as read from `.htaccess`, for some reason the "index.php"-less URLs still don't work ?

Are we sure all the correct Apache modules were enabled on that setup ? (mod_rewrite, mod_env)

**Comment by PVince81** [comment]
Ok, tried on a test instance and the URLs look fine on 9.0.1. URLs without index.php work so it is normal that share links are generated without index.php.

Now the question is why when updating to 9.0.3 the instance loses the index.php-less mode.

**Comment by PVince81** [comment]
We tried doing a manual tarball update on a CentOS docker instance and it worked fine, index.php-less mode was preserved.

So maybe something is different or wrong in the setup that does automated testing.

**Comment by SergioBertolinSG** [comment]
Ok, I think this is not an issue, but automated test problem needs to be addressed. @jvillafanez can you help with this? We used our centos 6 image manually and there weren't any problems with it.

**Comment by jvillafanez** [comment]
The errors I see from the tests are from 9.0.x to 9.1. Upgrades from 9.0.x to stable9 are working fine

**Comment by jvillafanez** [comment]
The contents of the .htaccess file after the upgrade to 9.1:
```
<IfModule mod_headers.c>
  <IfModule mod_setenvif.c>
    <IfModule mod_fcgid.c>
       SetEnvIfNoCase ^Authorization$ "(.+)" XAUTHORIZATION=$1
       RequestHeader set XAuthorization %{XAUTHORIZATION}e env=XAUTHORIZATION
    </IfModule>
    <IfModule mod_proxy_fcgi.c>
       SetEnvIfNoCase Authorization "(.+)" HTTP_AUTHORIZATION=$1
    </IfModule>
  </IfModule>

  <IfModule mod_env.c>
    # Add security and privacy related headers
    Header set X-Content-Type-Options "nosniff"
    Header set X-XSS-Protection "1; mode=block"
    Header set X-Robots-Tag "none"
    Header set X-Frame-Options "SAMEORIGIN"
    Header set X-Download-Options "noopen"
    Header set X-Permitted-Cross-Domain-Policies "none"
    SetEnv modHeadersAvailable true
  </IfModule>

  # Add cache control for CSS and JS files
  <FilesMatch "\.(css|js)$">
    Header set Cache-Control "max-age=7200, public"
  </FilesMatch>
</IfModule>
<IfModule mod_php5.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value always_populate_raw_post_data -1
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_php7.c>
  php_value upload_max_filesize 513M
  php_value post_max_size 513M
  php_value memory_limit 512M
  php_value mbstring.func_overload 0
  php_value default_charset 'UTF-8'
  php_value output_buffering 0
  <IfModule mod_env.c>
    SetEnv htaccessWorking true
  </IfModule>
</IfModule>
<IfModule mod_rewrite.c>
  RewriteEngine on
  RewriteRule .* - [env=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
  RewriteRule ^\.well-known/host-meta /public.php?service=host-meta [QSA,L]
  RewriteRule ^\.well-known/host-meta\.json /public.php?service=host-meta-json [QSA,L]
  RewriteRule ^\.well-known/carddav /remote.php/dav/ [R=301,L]
  RewriteRule ^\.well-known/caldav /remote.php/dav/ [R=301,L]
  RewriteRule ^remote/(.*) remote.php [QSA,L]
  RewriteRule ^(?:build|tests|config|lib|3rdparty|templates)/.* - [R=404,L]
  RewriteCond %{REQUEST_URI} !^/.well-known/acme-challenge/.*
  RewriteRule ^(?:\.|autotest|occ|issue|indie|db_|console).* - [R=404,L]
</IfModule>
<IfModule mod_mime.c>
  AddType image/svg+xml svg svgz
  AddEncoding gzip svgz
</IfModule>
<IfModule mod_dir.c>
  DirectoryIndex index.php index.html
</IfModule>
AddDefaultCharset utf-8
Options -Indexes
<IfModule pagespeed_module>
  ModPagespeed Off
</IfModule>
#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####

ErrorDocument 403 /core/templates/403.php
ErrorDocument 404 /core/templates/404.php
```

**Comment by SergioBertolinSG** [comment]
Still happening. (9.0.5 -> stable9.1)

**Comment by SergioBertolinSG** [comment]
This is fixed in 9.0.3. 9.0.X versions lesser than 9.0.3 will continue to fail. But I think it is enough to close this.

**Comment by PVince81** [comment]
Closing this as old. Please try with newer versions.
